Add Introductory Price to InAppBillingProduct

PhilippeCreytens opened the pull request:
Please take a moment to fill out the following:

Fixes #123  .

Changes Proposed in this pull request:
- Add introductory price to Android
- Add introductory price to iOS
- Implement boolean to check if an introductory price is available

PhilippeCreytens commented:
Hey James,
I have updated the pull request to only add Introductory price since iOS 11.2.
Regarding the android, yes this is available in the binding you have made, if an introductory price is available then this is part of the json that is being processed. An introductory price is available from the Google Play Store version 3.5, the earliest version now available that supports subscriptions.

marza91 commented:
Going to test this for our app, as we need this before we can do a release. Is there no way around needing iOS 11.2 to get these prices?
We are supposed to support iOS 8, but might have to reconsider if we can't get the prices to show up correctly.

jamesmontemagno commented:
@marza91 they added the api in 11.2 so it is impossible to have it before any release before that. 

There really isn't a reason to support anything older than iOS 10: https://data.apteligent.com/ios/

jamesmontemagno commented on src/Plugin.InAppBilling.iOS/InAppBillingImplementation.cs:
```diff
@@ -56,15 +56,18 @@ public InAppBillingImplementation()
+                Description = p.LocalizedDescription,
+                CurrencyCode = p.PriceLocale?.CurrencyCode ?? string.Empty,
+                LocalizedIntroductoryPrice = operatingSystemHasIntroductoryPrice && p.IntroductoryPrice != null ? p.IntroductoryPrice.LocalizedPrice() : "",
+                MicrosIntroductoryPrice = operatingSystemHasIntroductoryPrice && p.IntroductoryPrice != null ? (long)(p.Price.DoubleValue * 1000000d) : 0
```
shouldn't this be looking at p.IntroductoryPrice.DoubleValue?

jamesmontemagno replied:
maybe we should refactor this into a common method?

jamesmontemagno commented on src/Plugin.InAppBilling.iOS/InAppBillingImplementation.cs:
```diff
@@ -56,15 +56,18 @@ public InAppBillingImplementation()
-				Description = p.LocalizedDescription,
-				CurrencyCode = p.PriceLocale?.CurrencyCode ?? string.Empty
-			});
+            var operatingSystemHasIntroductoryPrice = NSProcessInfo.ProcessInfo.IsOperatingSystemAtLeastVersion(new NSOperatingSystemVersion(11, 2, 0));
```
This is wrong. Put it at the class level:

bool IsiOS112 => UIDevice.CurrentDevice.CheckSystemVersion (11,2);

jamesmontemagno commented:
Added a few more required tweaks to this one

PhilippeCreytens commented:
Hey James,

I have updated the PR, an extension method is now used to check whether an introduction price can be used and I have updated the DoubleValue to target the correct value.

jamesmontemagno commented:
Any changes or anything needed for UWP?

PhilippeCreytens commented:
No changes are needed for UWP, the InAppBillingProduct will say no IntroductoryPrice is available, according to the microsoft documentation I do not see any reference to an introductory price, only if the item is on sale or not but this is not the same as an introductory price.
